fence_token_storage.py

```diff
@@ -0,0 +1,156 @@
+    account credentials, i.e. "key_json."
+
+    This is abstracted as its own class so that we can stub out the Datastore dependency in tests.
+    """
+
+    def delete(self, fsa_key):
+        """
+        Delete the stored fence service account info for the fsa_key.
+        ":return returns the stored value for the key or None if it did not exist.
+        """
+        fence_service_account = fsa_key.get()
+        if fence_service_account:
+            fsa_key.delete()
+        return fence_service_account.key_json if fence_service_account else None
+
+    def get_or_create(self, fsa_key, prep_key_fn, fence_fetch_fn):
+        """Retrieve the stored for key, waiting as needed, or create and store the value for the key.
+
+        :param prep_key_fn The function to create the input to fence_fetch_fn from 'fsa_key' once we know
+        fence_fetch_fn will be called. This is separated from fence_fetch_fn so that this can be slow but not spend as
+         much time locking.
+        :param fence_fetch_fn: The function to fetch the credentials from the fence. This function will ensure that
+        fence_fetch_fn is not called multiple times concurrently. Arguments should work as
+        fence_fetch_fn(prep_key_fn(key)) -> returns the string fence account credentials.
+        :return returns the fence service account json key and the expiration time for that value.
+        """
+        fence_service_account = fsa_key.get()
+        now = datetime.datetime.now()
+        if fence_service_account is None or \
+                fence_service_account.expires_at is None or \
+                fence_service_account.expires_at < now:
+            fence_service_account = self._fetch_service_account(fsa_key, prep_key_fn, fence_fetch_fn)
+
+        return (fence_service_account.key_json, fence_service_account.expires_at)
+
+    def _fetch_service_account(self, fsa_key, prep_key_fn, fence_fetch_fn):
+        """
+        Fetch a new service account from fence. We must be careful that concurrent requests result in only one
+        key request to fence so the service account does not run out of keys (google limits to 10).
+        """
+        # Prep key before acquiring lock to keep lock duration as small as possible.
+        prepped_key = prep_key_fn(fsa_key)
+
+        if self._acquire_lock(fsa_key):
+            key_json = fence_fetch_fn(prepped_key)
+            fence_service_account = FenceServiceAccount(key_json=key_json,
+                                                        expires_at=datetime.datetime.now() + _FSA_KEY_LIFETIME,
+                                                        update_lock_timeout=None,
+                                                        key=fsa_key)
+            fence_service_account.put()
+
+        else:
+            fence_service_account = self._wait_for_update(fsa_key)
+            if not fence_service_account.expires_at or fence_service_account.expires_at < datetime.datetime.now():
```

Comment:
Change condition for ServiceAccountNotUpdatedException to raise that function if there is no expired_at message, i.e. it's still locked, instead of only if the updated account does not have a valid expires_at value.
Before, if the update lock expired during the _wait_for_update loop, we could return the FenceServiceAccount without a received key_json, as the update locked FenceServiceAccount does not have any value for expires_at. Now, the expired update locked value raises the exception.